rest/config.go

```diff
@@ -513,65 +515,32 @@ func (clusterConfig *ClusterConfig) GetCredentials() (string, string, string) {
 	return base.TransformBucketCredentials(clusterConfig.Username, clusterConfig.Password, *clusterConfig.Bucket)
 }
 
-// Reads a ServerConfig from raw data
-func ReadServerConfigFromData(runMode SyncGatewayRunMode, data []byte) (*ServerConfig, error) {
+// LoadServerConfig loads a ServerConfig from either a JSON file or from a URL
+func LoadServerConfig(runMode SyncGatewayRunMode, path string) (config *ServerConfig, err error) {
+	var dataReadCloser io.ReadCloser
 
-	data = base.ConvertBackQuotedStrings(data)
-	var config *ServerConfig
-	if err := json.Unmarshal(data, &config); err != nil {
-		return nil, err
-	}
-
-	config.RunMode = runMode
-
-	// Validation:
-	if err := config.setupAndValidateDatabases(); err != nil {
-		return nil, err
-	}
-
-	return config, nil
-}
-
-// Reads a ServerConfig from a URL.
-func ReadServerConfigFromUrl(runMode SyncGatewayRunMode, url string) (*ServerConfig, error) {
-
-	resp, err := http.Get(url)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
-	responseBody, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
-	}
-	return ReadServerConfigFromData(runMode, responseBody)
-
-}
-
-// Reads a ServerConfig from either a JSON file or from a URL.
-func ReadServerConfig(runMode SyncGatewayRunMode, path string) (*ServerConfig, error) {
 	if strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://") {
-		return ReadServerConfigFromUrl(runMode, path)
+		resp, err := http.Get(path)
+		if err != nil {
+			return nil, err
+		} else if resp.StatusCode >= 300 {
+			return nil, base.HTTPErrorf(resp.StatusCode, http.StatusText(resp.StatusCode))
+		}
+		dataReadCloser = resp.Body
 	} else {
-		return ReadServerConfigFromFile(runMode, path)
+		dataReadCloser, err = os.Open(path)
+		if err != nil {
+			return nil, err
+		}
 	}
-}
 
-// Reads a ServerConfig from a JSON file.
-func ReadServerConfigFromFile(runMode SyncGatewayRunMode, path string) (*ServerConfig, error) {
-	file, err := os.Open(path)
-	if err != nil {
-		return nil, err
-	}
-	defer file.Close()
+	defer dataReadCloser.Close()
+	return readServerConfig(runMode, dataReadCloser)
+}
 
-	data, err := ioutil.ReadAll(file)
-	if err != nil {
-		return nil, err
-	}
-	data = base.ConvertBackQuotedStrings(data)
-	var config *ServerConfig
-	if err := json.Unmarshal(data, &config); err != nil {
+// readServerConfig returns a validated ServerConfig from an io.Reader
+func readServerConfig(runMode SyncGatewayRunMode, r io.Reader) (config *ServerConfig, err error) {
+	if err := decodeAndSanitiseConfig(r, config); err != nil {
```

Comment:
As discussed, add unit test and check whether we should add a more user-friendly wrapper to the error returned by decode when invalid keys are present.

Reply:
Might be nice to have the test run all our example configs through this, as it will prevent future regressions in those examples.

Reply:
> As discussed, add unit test and check whether we should add a more user-friendly wrapper to the error returned by decode when invalid keys are present.

SG wraps errors coming from LoadServerConfig with this on startup:
```
base.Fatalf(base.KeyAll, "Error reading config file %s: %v", base.UD(filename), err)
```

An invalid key would look like this:

```
2019-02-22T16:57:08.188Z [ERR] Error reading config file sg_config.json: json: unknown field "num_index_replicaas" -- rest.ParseCommandLine() at config.go:747
```